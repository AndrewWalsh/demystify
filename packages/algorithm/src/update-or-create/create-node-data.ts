import decodeUriComponent from "decode-uri-component";
import qs from "fast-querystring";
import { convertXML } from "simple-xml-to-json";
import { HarRestful, JSONType, NodeData } from "../types/index.js";
import {
  createSchemaElseUndefined,
  parseJSON,
  getAuthHeaders,
  filterIgnoreHeaders,
} from "../utils/index.js";
import { ValidHar } from "../har-to-hartype/index.js";
import { Schema } from "genson-js/dist/types.js";
import { Cookie } from "har-format";

const parseRequestBody = (harRequest: HarRestful): JSONType => {
  const { mimeType, text } = harRequest.request.postData || {};
  if (mimeType?.startsWith("application/x-www-form-urlencoded") && text)
    return qs.parse(text) as JSONType;
  if (mimeType?.startsWith("application/json")) return parseJSON(text);
  return null;
};

const parseResponseBody = (harResponse: HarRestful): JSONType => {
  const { mimeType, text } = harResponse.response.content;
  if (mimeType?.startsWith("application/json")) return parseJSON(text);
  return null;
};

const expectFilterToContain = ["token", "session"];
const filterCookies = (cookies: Cookie[]) => {
  return cookies.reduce((acc, { name }) => {
    if (expectFilterToContain.some(filter => name.toLowerCase().includes(filter))) {
      return { ...acc, [name]: '' };
    }
    return acc;
  }, {});
};

export const createNodeData = ({ data: { har, kind } }: { data: ValidHar }): NodeData => {
  // 0 - Parse request/response bodies in relation to mime type
  let requestBody: JSONType | undefined;
  let responseBody: JSONType | undefined;
  let requestBodySchema: Schema | undefined;
  let responseBodySchema: Schema | undefined;
  if (kind === "rest-xml") {
    requestBody = har.request.postData?.text;
    responseBody = har.response.content.text;
    requestBodySchema = requestBody
      ? createSchemaElseUndefined(convertXML(requestBody))
      : undefined;
    responseBodySchema = responseBody
      ? createSchemaElseUndefined(convertXML(responseBody))
      : undefined;
  } else {
    requestBody = parseRequestBody(har);
    responseBody = parseResponseBody(har);
    requestBodySchema = createSchemaElseUndefined(requestBody);
    responseBodySchema = createSchemaElseUndefined(responseBody);
  }
  // 1 - Define node data
  const requestHeaders = getAuthHeaders(har.request.headers);
  const responseHeaders = filterIgnoreHeaders(har.response.headers).map(({ name }) => name);
  const method = har.request.method;
  const pathname = decodeUriComponent(new URL(har.request.url).pathname);
  const requestMime = har.request.postData?.mimeType;
  const responseMime = har.response.content.mimeType;
  const statusCode = har.response.status.toString();
  const cookies = har.request.cookies;
  const protocol = new URL(har.request.url).protocol as "http:" | "https:";
  return {
    protocol,
    mostRecentPathname: pathname,
    methods: {
      [method]: {
        ...(cookies && cookies.length && { cookies: filterCookies(har.request.cookies) }),
        ...(requestMime && {
          request: {
            [requestMime]: {
              kind,
              body: requestBodySchema,
              mostRecent: requestBody,
            },
          },
        }),
        requestHeaders: requestHeaders,
        response: {
          [statusCode]: {
            [responseMime]: {
              kind,
              body: responseBodySchema,
              mostRecent: responseBody,
            },
          },
        },
        responseHeaders: responseHeaders,
        ...(har.request.queryString?.length && { queryParameters: har.request.queryString.reduce((acc, cur) => ({ ...acc, [cur.name]: cur.value }), {}) }),
      },
    },
  };
};
