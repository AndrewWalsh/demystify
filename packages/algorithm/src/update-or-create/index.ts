export * from './update-or-create.js'
