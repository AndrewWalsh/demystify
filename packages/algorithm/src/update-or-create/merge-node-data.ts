import { mergeSchemas, Schema } from "genson-js";
import { NodeData } from "../types/index.js";

type Data = NodeData["methods"]["get"];
type Req = NonNullable<NodeData["methods"]["get"]["request"]>;
type Res = NodeData["methods"]["get"]["response"];
type Cookies = NodeData["methods"]["get"]["cookies"];

const mergeCookies = (dest: NonNullable<Cookies>, src: NonNullable<Cookies>) => ({
  ...dest,
  ...src,
});

const mergeRequest = (dest: Data, src: Req = {}) => {
  Object.entries(src).forEach(([mediaType, srcData]) => {
    // Nothing in dest or src to merge
    if (!srcData || !srcData.body) return;
    dest["request"] ??= {};
    // Nothing in dest to merge, but src has data
    if (!dest["request"][mediaType]?.body) {
      dest["request"][mediaType] = srcData;
      return;
    }
    // Merge schemas
    dest["request"][mediaType].body = mergeSchemas([
      dest["request"][mediaType].body,
      srcData.body,
    ]);
    dest["request"][mediaType].mostRecent = srcData.mostRecent;
  });
};

const mergeResponse = (dest: Data, src: Res = {}) => {
  Object.entries(src).forEach(([statusCode, srcMediaTypeObj]) => {
    // statusCode in src does not exist in dest
    if (!dest["response"]?.[statusCode]) {
      dest["response"][statusCode] = srcMediaTypeObj;
      return;
    }
    // statusCode exists in both dest and src
    Object.entries(srcMediaTypeObj).forEach(([mediaType, mediaTypeData]) => {
      const srcData = srcMediaTypeObj[mediaType]!;
      if (!dest["response"]?.[statusCode]?.[mediaType]) {
        // dest does not contain mediaType, set from src
        dest["response"][statusCode]![mediaType] = srcData;
        return;
      } else {
        // update or create schemas
        const selector = dest["response"][statusCode]![mediaType]!;
        if (!selector.body) {
          selector.body = srcData.body;
        } else if (srcData.body) {
          selector.body = mergeSchemas([selector.body, srcData.body]);
        }
        selector.mostRecent =
          mediaTypeData.mostRecent;
      }
    });
  });
};

export const mergeNodeData = (dest: NodeData, src: NodeData): NodeData => {
  dest.protocol = src.protocol;
  dest.mostRecentPathname = src.mostRecentPathname;
  for (const [method, methodObj] of Object.entries(src.methods)) {
    // Method doesn't exist in dest, set src and continue to next method
    if (!dest.methods[method]) {
      dest.methods[method] = methodObj;
      continue;
    }

    const srcSchema = src.methods[method]!;
    const destSchema = dest.methods[method]!;
    // Merge cookies
    if (src.methods[method]?.cookies) {
      if (!dest.methods[method]?.cookies) {
        dest.methods[method].cookies = src.methods[method].cookies;
      } else {
        dest.methods[method].cookies = mergeCookies(
          dest.methods[method].cookies,
          src.methods[method].cookies
        );
      }
    }
    // Merge request
    if (destSchema.request || srcSchema.request) {
      mergeRequest(destSchema, srcSchema.request);
    }
    // Merge query params
    if (destSchema.queryParameters || srcSchema.queryParameters) {
      destSchema.queryParameters = {
        ...destSchema.queryParameters,
        ...srcSchema.queryParameters
      }
    }
    // Merge request headers
    if (destSchema.requestHeaders || srcSchema.requestHeaders) {
      const destReqHeaders = destSchema.requestHeaders || [];
      const srcReqHeaders = srcSchema.requestHeaders || [];
      destSchema.requestHeaders = Array.from(
        new Set([...destReqHeaders, ...srcReqHeaders])
      );
    }
    // Merge response headers
    if (destSchema.responseHeaders || srcSchema.responseHeaders) {
      const destResHeaders = destSchema.responseHeaders || [];
      const srcResHeaders = srcSchema.responseHeaders || [];
      destSchema.responseHeaders = Array.from(
        new Set([...destResHeaders, ...srcResHeaders])
      );
    }
    // Merge responses
    mergeResponse(destSchema, methodObj.response);
  }
  return dest;
};
