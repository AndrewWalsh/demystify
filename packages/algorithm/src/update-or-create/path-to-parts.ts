import { HarRestJson, HarRestXml } from '../types/index.js';

export const pathToParts = (harEntry: HarRestJson | HarRestXml): string[] => {
  const url = new URL(harEntry.request.url);
  return url.pathname
    .split('/')
    .filter(part => part.length > 0);
};
