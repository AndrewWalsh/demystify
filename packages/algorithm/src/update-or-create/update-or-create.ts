import { IrNode } from '../types/index.js';
import { pathToParts } from './path-to-parts.js';
import { createNodeData } from './create-node-data.js';
import { createNode } from '../utils/index.js';
import { ValidHar } from '../har-to-hartype/index.js';
import { mergeNodeData } from './merge-node-data.js';
import { getStaticChildMatch } from '../utils/index.js';
import { matchDynamicChildren } from '../utils/dynamic-children-helpers.js';

type UpdateOrCreateNodeParams = {
  node: IrNode;
  parts: string[];
  data: ValidHar;
}

function updateOrCreateNode({ node, parts, data }: UpdateOrCreateNodeParams): IrNode {
  // Base case, no more parts to match. Update or create the node
  // If node.data is null, then it becomes the harEntry (create)
  // Otherwise, merge the harEntry into the existing data (update)
  if (parts.length === 0) {
    const src = createNodeData({ data });
    const dest = node.data;
    if (!dest) node.data = src;
    else node.data = mergeNodeData(dest, src);
    return node;
  }
  const part = parts[0]!;
  const dynamicMatch = matchDynamicChildren(part, node.childrenDynamic);
  // Look at dynamic nodes first
  // This is important because we want to match against this first, should it exist
  // Prior to going down that path, confirm that an endpoint exists in the subgraph
  if (dynamicMatch && nodeHasDataForPath(dynamicMatch, parts.slice(1))) {
    return updateOrCreateNode({ node: dynamicMatch, parts: parts.slice(1), data });
  }
  // Look for an existing match, if found, continue the search
  const staticMatch = getStaticChildMatch(part, node);
  if (staticMatch) {
    return updateOrCreateNode({ node: staticMatch, parts: parts.slice(1), data });
  }
  // Create a node if no match is found
  // If we get to this point, parts is non-empty and there is neither a static or dynamic match
  node.childrenStatic ??= {};
  node.childrenStatic[part] = createNode({ key: part, parent: node });
  return updateOrCreateNode({ node: node.childrenStatic[part], parts: parts.slice(1), data });
}

const nodeHasDataForPath = (node: IrNode, parts: string[]): boolean => {
  if (parts.length === 0) {
    if (node.data) {
      return true;
    }
    return false;
  }
  const part = parts[0]!;
  if (node.childrenStatic && node.childrenStatic[part]) {
    if (nodeHasDataForPath(node.childrenStatic[part], parts.slice(1))) {
      return true;
    }
  }
  for (const dynamicNode of node.childrenDynamic) {
    if (nodeHasDataForPath(dynamicNode, parts.slice(1))) {
      return true;
    }
  }
  return false;
};

type UpdateOrCreateParams = {
  data: ValidHar;
  rootNode: IrNode | null;
}
export function updateOrCreate({ data, rootNode }: UpdateOrCreateParams): { root: IrNode, inserted: IrNode } {
  const { har } = data;
  if (!rootNode) {
    const host = new URL(har.request.url).host;
    rootNode = createNode({ key: host });
  }
  const newNode = updateOrCreateNode({
    node: rootNode,
    parts: pathToParts(har),
    data,
  });
  return { root: rootNode, inserted: newNode };
}
