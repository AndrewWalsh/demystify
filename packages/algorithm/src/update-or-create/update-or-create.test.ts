import { describe, it, expect } from "vitest";
import { updateOrCreate } from "./update-or-create.js";
import { HarRestJson } from "../types/index.js";
import { createNode } from "../utils/index.js";
import restPostHar from "../__fixtures__/rest-post.json" with { type: "json" };
import restPostHarAlt from "../__fixtures__/rest-post-alt.json" with { type: "json" };
import xmlHar from "../__fixtures__/xml.json" with { type: "json" };
import { ValidHar } from "../har-to-hartype/index.js";

describe("findOrCreate REST JSON", () => {
  const data: ValidHar = {
    har: restPostHar.log.entries[0] as HarRestJson,
    kind: "rest-json",
  };
  it("when rootNode is not provided, return a new root node matching snapshot", () => {
    const result = updateOrCreate({ data, rootNode: null });
    expect(result.root).toMatchSnapshot();
  });

  it("operations are idempotent", () => {
    const rootNode = createNode({ key: "example.com" });
    const result = updateOrCreate({ data, rootNode });
    const foundNode = updateOrCreate({ data, rootNode: result.root });
    expect(foundNode).toEqual(result);
  });
});

describe("findOrCreate REST JSON alternative", () => {
  const dataOne: ValidHar = {
    har: restPostHarAlt.log.entries[0] as HarRestJson,
    kind: "rest-json",
  };
  const dataTwo: ValidHar = {
    har: restPostHar.log.entries[0] as HarRestJson,
    kind: "rest-json",
  };
  it("when rootNode is not provided, return a new root node matching snapshot", () => {
    const result = updateOrCreate({ data: dataOne, rootNode: null });
    expect(result.root).toMatchSnapshot();
  });

  it("merges harEntry into an existing node", () => {
    const resultOne = updateOrCreate({ data: dataOne, rootNode: null });
    const resultTwo = updateOrCreate({ data: dataTwo, rootNode: resultOne.root });
    expect(resultTwo.root).toMatchSnapshot();
  });
});

describe("findOrCreate XML", () => {
  const data: ValidHar = {
    har: xmlHar.log.entries[0] as HarRestJson,
    kind: "rest-xml",
  };
  it("when rootNode is not provided, return a new root node matching snapshot", () => {
    const result = updateOrCreate({ data, rootNode: null });
    expect(result.root).toMatchSnapshot();
  });

  it("merges data into an existing node", () => {
    const resultOne = updateOrCreate({ data, rootNode: null });
    const resultTwo = updateOrCreate({ data, rootNode: resultOne.root });
    expect(resultTwo.root).toMatchSnapshot();
  });

  it("operations are idempotent", () => {
    const rootNode = createNode({ key: "example.com" });
    const result = updateOrCreate({ data, rootNode });
    const foundNode = updateOrCreate({ data, rootNode: result.root });
    expect(foundNode).toEqual(result);
  });
});

describe("updateOrCreate Additional Cases", () => {
  it("handles an empty path in the HAR request", () => {
    const data: ValidHar = {
      har: {
        request: {
          method: "GET",
          url: "http://example.com",
          headers: [],
          queryString: [],
          cookies: [],
          httpVersion: "",
          bodySize: 0,
          headersSize: 0,
        },
        response: {
          status: 200,
          content: { mimeType: "application/json", text: "" },
          headers: [],
          cookies: [],
          redirectURL: "",
          headersSize: 0,
          bodySize: 0,
        },
        startedDateTime: "",
        time: 0,
      } as unknown as HarRestJson,
      kind: "rest-json",
    };
    const result = updateOrCreate({ data, rootNode: null });
    expect(result).toBeDefined();
  });

  it("merges multiple times with no path changes", () => {
    const data: ValidHar = {
      har: restPostHar.log.entries[0] as HarRestJson,
      kind: "rest-json",
    };
    const firstUpdate = updateOrCreate({ data, rootNode: null });
    const secondUpdate = updateOrCreate({ data, rootNode: firstUpdate.root });
    const thirdUpdate = updateOrCreate({ data, rootNode: secondUpdate.root });
    expect(thirdUpdate.root).toEqual(secondUpdate.root);
  });

  it("handles missing response content gracefully", () => {
    const data: ValidHar = {
      har: {
        ...restPostHar.log.entries[0],
        response: {
          ...restPostHar.log.entries[0]!.response,
          content: { mimeType: "", text: "" },
        },
      } as HarRestJson,
      kind: "rest-json",
    };
    const result = updateOrCreate({ data, rootNode: null });
    expect(result).toBeDefined();
  });
});
