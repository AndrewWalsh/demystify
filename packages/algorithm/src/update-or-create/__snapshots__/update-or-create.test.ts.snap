// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`findOrCreate REST JSON > when rootNode is not provided, return a new root node matching snapshot 1`] = `
{
  "childrenDynamic": [],
  "childrenStatic": {
    "v1": {
      "childrenDynamic": [],
      "childrenStatic": {
        "pets": {
          "childrenDynamic": [],
          "childrenStatic": {},
          "data": {
            "methods": {
              "POST": {
                "request": {
                  "application/json": {
                    "body": {
                      "properties": {
                        "age": {
                          "type": "integer",
                        },
                        "name": {
                          "type": "string",
                        },
                        "tags": {
                          "items": {
                            "type": "string",
                          },
                          "type": "array",
                        },
                        "type": {
                          "type": "string",
                        },
                        "vaccinated": {
                          "type": "boolean",
                        },
                      },
                      "required": [
                        "name",
                        "type",
                        "age",
                        "vaccinated",
                        "tags",
                      ],
                      "type": "object",
                    },
                    "kind": "rest-json",
                    "mostRecent": {
                      "age": 3,
                      "name": "Fluffy",
                      "tags": [
                        "friendly",
                        "indoor",
                      ],
                      "type": "cat",
                      "vaccinated": true,
                    },
                  },
                },
                "requestHeaders": [],
                "response": {
                  "201": {
                    "application/json": {
                      "body": {
                        "properties": {
                          "age": {
                            "type": "integer",
                          },
                          "created_at": {
                            "type": "string",
                          },
                          "id": {
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "tags": {
                            "items": {
                              "type": "string",
                            },
                            "type": "array",
                          },
                          "type": {
                            "type": "string",
                          },
                          "vaccinated": {
                            "type": "boolean",
                          },
                        },
                        "required": [
                          "id",
                          "name",
                          "type",
                          "age",
                          "vaccinated",
                          "tags",
                          "created_at",
                        ],
                        "type": "object",
                      },
                      "kind": "rest-json",
                      "mostRecent": {
                        "age": 3,
                        "created_at": "2023-05-20T10:30:15.123Z",
                        "id": 12345,
                        "name": "Fluffy",
                        "tags": [
                          "friendly",
                          "indoor",
                        ],
                        "type": "cat",
                        "vaccinated": true,
                      },
                    },
                  },
                },
                "responseHeaders": [],
              },
            },
            "mostRecentPathname": "/v1/pets",
            "protocol": "https:",
          },
          "key": "pets",
          "parent": [Circular],
        },
      },
      "data": null,
      "key": "v1",
      "parent": [Circular],
    },
  },
  "data": null,
  "key": "api.example.com",
  "parent": null,
}
`;

exports[`findOrCreate REST JSON alternative > merges harEntry into an existing node 1`] = `
{
  "childrenDynamic": [],
  "childrenStatic": {
    "v1": {
      "childrenDynamic": [],
      "childrenStatic": {
        "pets": {
          "childrenDynamic": [],
          "childrenStatic": {},
          "data": {
            "methods": {
              "POST": {
                "request": {
                  "application/json": {
                    "body": {
                      "properties": {
                        "age": {
                          "type": "integer",
                        },
                        "name": {
                          "type": "string",
                        },
                        "tags": {
                          "items": {
                            "type": "string",
                          },
                          "type": "array",
                        },
                        "type": {
                          "type": "string",
                        },
                        "vaccinated": {
                          "type": "boolean",
                        },
                      },
                      "required": [
                        "name",
                        "type",
                        "age",
                        "vaccinated",
                        "tags",
                      ],
                      "type": "object",
                    },
                    "kind": "rest-json",
                    "mostRecent": {
                      "age": 3,
                      "name": "Fluffy",
                      "tags": [
                        "friendly",
                        "indoor",
                      ],
                      "type": "cat",
                      "vaccinated": true,
                    },
                  },
                },
                "requestHeaders": [],
                "response": {
                  "201": {
                    "application/json": {
                      "body": {
                        "properties": {
                          "age": {
                            "type": "integer",
                          },
                          "created_at": {
                            "type": "string",
                          },
                          "id": {
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "tags": {
                            "items": {
                              "type": "string",
                            },
                            "type": "array",
                          },
                          "type": {
                            "type": "string",
                          },
                          "vaccinated": {
                            "type": "boolean",
                          },
                        },
                        "required": [
                          "id",
                          "name",
                          "type",
                          "age",
                          "vaccinated",
                          "tags",
                          "created_at",
                        ],
                        "type": "object",
                      },
                      "kind": "rest-json",
                      "mostRecent": {
                        "age": 3,
                        "created_at": "2023-05-20T10:30:15.123Z",
                        "id": 12345,
                        "name": "Fluffy",
                        "tags": [
                          "friendly",
                          "indoor",
                        ],
                        "type": "cat",
                        "vaccinated": true,
                      },
                    },
                  },
                },
                "responseHeaders": [],
              },
            },
            "mostRecentPathname": "/v1/pets",
            "protocol": "https:",
          },
          "key": "pets",
          "parent": [Circular],
        },
        "track": {
          "childrenDynamic": [],
          "childrenStatic": {},
          "data": {
            "methods": {
              "POST": {
                "cookies": {},
                "queryParameters": {
                  "alt": "json",
                  "key": "abc",
                },
                "request": {
                  "application/json": {
                    "body": {
                      "properties": {
                        "test": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "test",
                      ],
                      "type": "object",
                    },
                    "kind": "rest-json",
                    "mostRecent": {
                      "test": "integer",
                    },
                  },
                },
                "requestHeaders": [],
                "response": {
                  "200": {
                    "application/json": {
                      "body": {
                        "properties": {
                          "test": {
                            "type": "integer",
                          },
                        },
                        "required": [
                          "test",
                        ],
                        "type": "object",
                      },
                      "kind": "rest-json",
                      "mostRecent": {
                        "test": 1,
                      },
                    },
                  },
                },
                "responseHeaders": [],
              },
            },
            "mostRecentPathname": "/v1/track",
            "protocol": "https:",
          },
          "key": "track",
          "parent": [Circular],
        },
      },
      "data": null,
      "key": "v1",
      "parent": [Circular],
    },
  },
  "data": null,
  "key": "www.example.com",
  "parent": null,
}
`;

exports[`findOrCreate REST JSON alternative > when rootNode is not provided, return a new root node matching snapshot 1`] = `
{
  "childrenDynamic": [],
  "childrenStatic": {
    "v1": {
      "childrenDynamic": [],
      "childrenStatic": {
        "track": {
          "childrenDynamic": [],
          "childrenStatic": {},
          "data": {
            "methods": {
              "POST": {
                "cookies": {},
                "queryParameters": {
                  "alt": "json",
                  "key": "abc",
                },
                "request": {
                  "application/json": {
                    "body": {
                      "properties": {
                        "test": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "test",
                      ],
                      "type": "object",
                    },
                    "kind": "rest-json",
                    "mostRecent": {
                      "test": "integer",
                    },
                  },
                },
                "requestHeaders": [],
                "response": {
                  "200": {
                    "application/json": {
                      "body": {
                        "properties": {
                          "test": {
                            "type": "integer",
                          },
                        },
                        "required": [
                          "test",
                        ],
                        "type": "object",
                      },
                      "kind": "rest-json",
                      "mostRecent": {
                        "test": 1,
                      },
                    },
                  },
                },
                "responseHeaders": [],
              },
            },
            "mostRecentPathname": "/v1/track",
            "protocol": "https:",
          },
          "key": "track",
          "parent": [Circular],
        },
      },
      "data": null,
      "key": "v1",
      "parent": [Circular],
    },
  },
  "data": null,
  "key": "www.example.com",
  "parent": null,
}
`;

exports[`findOrCreate XML > merges data into an existing node 1`] = `
{
  "childrenDynamic": [],
  "childrenStatic": {
    "v1": {
      "childrenDynamic": [],
      "childrenStatic": {
        "pets": {
          "childrenDynamic": [],
          "childrenStatic": {},
          "data": {
            "methods": {
              "POST": {
                "request": {
                  "application/xml": {
                    "body": {
                      "properties": {
                        "pet": {
                          "properties": {
                            "children": {
                              "items": {
                                "properties": {
                                  "age": {
                                    "properties": {
                                      "content": {
                                        "type": "string",
                                      },
                                    },
                                    "required": [
                                      "content",
                                    ],
                                    "type": "object",
                                  },
                                  "name": {
                                    "properties": {
                                      "content": {
                                        "type": "string",
                                      },
                                    },
                                    "required": [
                                      "content",
                                    ],
                                    "type": "object",
                                  },
                                  "tags": {
                                    "properties": {
                                      "children": {
                                        "items": {
                                          "properties": {
                                            "tag": {
                                              "properties": {
                                                "content": {
                                                  "type": "string",
                                                },
                                              },
                                              "required": [
                                                "content",
                                              ],
                                              "type": "object",
                                            },
                                          },
                                          "required": [
                                            "tag",
                                          ],
                                          "type": "object",
                                        },
                                        "type": "array",
                                      },
                                    },
                                    "required": [
                                      "children",
                                    ],
                                    "type": "object",
                                  },
                                  "type": {
                                    "properties": {
                                      "content": {
                                        "type": "string",
                                      },
                                    },
                                    "required": [
                                      "content",
                                    ],
                                    "type": "object",
                                  },
                                  "vaccinated": {
                                    "properties": {
                                      "content": {
                                        "type": "string",
                                      },
                                    },
                                    "required": [
                                      "content",
                                    ],
                                    "type": "object",
                                  },
                                },
                                "type": "object",
                              },
                              "type": "array",
                            },
                          },
                          "required": [
                            "children",
                          ],
                          "type": "object",
                        },
                      },
                      "required": [
                        "pet",
                      ],
                      "type": "object",
                    },
                    "kind": "rest-xml",
                    "mostRecent": "<?xml version="1.0" encoding="UTF-8"?><pet><name>Fluffy</name><type>cat</type><age>3</age><vaccinated>true</vaccinated><tags><tag>friendly</tag><tag>indoor</tag></tags></pet>",
                  },
                },
                "requestHeaders": [],
                "response": {
                  "201": {
                    "application/xml": {
                      "body": {
                        "properties": {
                          "pet": {
                            "properties": {
                              "children": {
                                "items": {
                                  "properties": {
                                    "age": {
                                      "properties": {
                                        "content": {
                                          "type": "string",
                                        },
                                      },
                                      "required": [
                                        "content",
                                      ],
                                      "type": "object",
                                    },
                                    "created_at": {
                                      "properties": {
                                        "content": {
                                          "type": "string",
                                        },
                                      },
                                      "required": [
                                        "content",
                                      ],
                                      "type": "object",
                                    },
                                    "id": {
                                      "properties": {
                                        "content": {
                                          "type": "string",
                                        },
                                      },
                                      "required": [
                                        "content",
                                      ],
                                      "type": "object",
                                    },
                                    "name": {
                                      "properties": {
                                        "content": {
                                          "type": "string",
                                        },
                                      },
                                      "required": [
                                        "content",
                                      ],
                                      "type": "object",
                                    },
                                    "tags": {
                                      "properties": {
                                        "children": {
                                          "items": {
                                            "properties": {
                                              "tag": {
                                                "properties": {
                                                  "content": {
                                                    "type": "string",
                                                  },
                                                },
                                                "required": [
                                                  "content",
                                                ],
                                                "type": "object",
                                              },
                                            },
                                            "required": [
                                              "tag",
                                            ],
                                            "type": "object",
                                          },
                                          "type": "array",
                                        },
                                      },
                                      "required": [
                                        "children",
                                      ],
                                      "type": "object",
                                    },
                                    "type": {
                                      "properties": {
                                        "content": {
                                          "type": "string",
                                        },
                                      },
                                      "required": [
                                        "content",
                                      ],
                                      "type": "object",
                                    },
                                    "vaccinated": {
                                      "properties": {
                                        "content": {
                                          "type": "string",
                                        },
                                      },
                                      "required": [
                                        "content",
                                      ],
                                      "type": "object",
                                    },
                                  },
                                  "type": "object",
                                },
                                "type": "array",
                              },
                            },
                            "required": [
                              "children",
                            ],
                            "type": "object",
                          },
                        },
                        "required": [
                          "pet",
                        ],
                        "type": "object",
                      },
                      "kind": "rest-xml",
                      "mostRecent": "<?xml version="1.0" encoding="UTF-8"?><pet><id>12345</id><name>Fluffy</name><type>cat</type><age>3</age><vaccinated>true</vaccinated><tags><tag>friendly</tag><tag>indoor</tag></tags><created_at>2023-05-20T10:30:15.123Z</created_at></pet>",
                    },
                  },
                },
                "responseHeaders": [],
              },
            },
            "mostRecentPathname": "/v1/pets",
            "protocol": "https:",
          },
          "key": "pets",
          "parent": [Circular],
        },
      },
      "data": null,
      "key": "v1",
      "parent": [Circular],
    },
  },
  "data": null,
  "key": "api.example.com",
  "parent": null,
}
`;

exports[`findOrCreate XML > when rootNode is not provided, return a new root node matching snapshot 1`] = `
{
  "childrenDynamic": [],
  "childrenStatic": {
    "v1": {
      "childrenDynamic": [],
      "childrenStatic": {
        "pets": {
          "childrenDynamic": [],
          "childrenStatic": {},
          "data": {
            "methods": {
              "POST": {
                "request": {
                  "application/xml": {
                    "body": {
                      "properties": {
                        "pet": {
                          "properties": {
                            "children": {
                              "items": {
                                "properties": {
                                  "age": {
                                    "properties": {
                                      "content": {
                                        "type": "string",
                                      },
                                    },
                                    "required": [
                                      "content",
                                    ],
                                    "type": "object",
                                  },
                                  "name": {
                                    "properties": {
                                      "content": {
                                        "type": "string",
                                      },
                                    },
                                    "required": [
                                      "content",
                                    ],
                                    "type": "object",
                                  },
                                  "tags": {
                                    "properties": {
                                      "children": {
                                        "items": {
                                          "properties": {
                                            "tag": {
                                              "properties": {
                                                "content": {
                                                  "type": "string",
                                                },
                                              },
                                              "required": [
                                                "content",
                                              ],
                                              "type": "object",
                                            },
                                          },
                                          "required": [
                                            "tag",
                                          ],
                                          "type": "object",
                                        },
                                        "type": "array",
                                      },
                                    },
                                    "required": [
                                      "children",
                                    ],
                                    "type": "object",
                                  },
                                  "type": {
                                    "properties": {
                                      "content": {
                                        "type": "string",
                                      },
                                    },
                                    "required": [
                                      "content",
                                    ],
                                    "type": "object",
                                  },
                                  "vaccinated": {
                                    "properties": {
                                      "content": {
                                        "type": "string",
                                      },
                                    },
                                    "required": [
                                      "content",
                                    ],
                                    "type": "object",
                                  },
                                },
                                "type": "object",
                              },
                              "type": "array",
                            },
                          },
                          "required": [
                            "children",
                          ],
                          "type": "object",
                        },
                      },
                      "required": [
                        "pet",
                      ],
                      "type": "object",
                    },
                    "kind": "rest-xml",
                    "mostRecent": "<?xml version="1.0" encoding="UTF-8"?><pet><name>Fluffy</name><type>cat</type><age>3</age><vaccinated>true</vaccinated><tags><tag>friendly</tag><tag>indoor</tag></tags></pet>",
                  },
                },
                "requestHeaders": [],
                "response": {
                  "201": {
                    "application/xml": {
                      "body": {
                        "properties": {
                          "pet": {
                            "properties": {
                              "children": {
                                "items": {
                                  "properties": {
                                    "age": {
                                      "properties": {
                                        "content": {
                                          "type": "string",
                                        },
                                      },
                                      "required": [
                                        "content",
                                      ],
                                      "type": "object",
                                    },
                                    "created_at": {
                                      "properties": {
                                        "content": {
                                          "type": "string",
                                        },
                                      },
                                      "required": [
                                        "content",
                                      ],
                                      "type": "object",
                                    },
                                    "id": {
                                      "properties": {
                                        "content": {
                                          "type": "string",
                                        },
                                      },
                                      "required": [
                                        "content",
                                      ],
                                      "type": "object",
                                    },
                                    "name": {
                                      "properties": {
                                        "content": {
                                          "type": "string",
                                        },
                                      },
                                      "required": [
                                        "content",
                                      ],
                                      "type": "object",
                                    },
                                    "tags": {
                                      "properties": {
                                        "children": {
                                          "items": {
                                            "properties": {
                                              "tag": {
                                                "properties": {
                                                  "content": {
                                                    "type": "string",
                                                  },
                                                },
                                                "required": [
                                                  "content",
                                                ],
                                                "type": "object",
                                              },
                                            },
                                            "required": [
                                              "tag",
                                            ],
                                            "type": "object",
                                          },
                                          "type": "array",
                                        },
                                      },
                                      "required": [
                                        "children",
                                      ],
                                      "type": "object",
                                    },
                                    "type": {
                                      "properties": {
                                        "content": {
                                          "type": "string",
                                        },
                                      },
                                      "required": [
                                        "content",
                                      ],
                                      "type": "object",
                                    },
                                    "vaccinated": {
                                      "properties": {
                                        "content": {
                                          "type": "string",
                                        },
                                      },
                                      "required": [
                                        "content",
                                      ],
                                      "type": "object",
                                    },
                                  },
                                  "type": "object",
                                },
                                "type": "array",
                              },
                            },
                            "required": [
                              "children",
                            ],
                            "type": "object",
                          },
                        },
                        "required": [
                          "pet",
                        ],
                        "type": "object",
                      },
                      "kind": "rest-xml",
                      "mostRecent": "<?xml version="1.0" encoding="UTF-8"?><pet><id>12345</id><name>Fluffy</name><type>cat</type><age>3</age><vaccinated>true</vaccinated><tags><tag>friendly</tag><tag>indoor</tag></tags><created_at>2023-05-20T10:30:15.123Z</created_at></pet>",
                    },
                  },
                },
                "responseHeaders": [],
              },
            },
            "mostRecentPathname": "/v1/pets",
            "protocol": "https:",
          },
          "key": "pets",
          "parent": [Circular],
        },
      },
      "data": null,
      "key": "v1",
      "parent": [Circular],
    },
  },
  "data": null,
  "key": "api.example.com",
  "parent": null,
}
`;
